File: Data Structures & Algorithms/binary-tree-diameter/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameter = 0;
    diameterOfBinaryTree(root) {
        this.height(root);
        return this.diameter
    }
    height(root) {
        if (!root){
            return 0
        }
        // Get the height of left and right subtrees
        const leftHeight = this.height(root.left);
        const rightHeight = this.height(root.right);

        // Update diameter (longest path through this node)
        this.diameter = Math.max(this.diameter, leftHeight + rightHeight);

        // Return the height of the current subtree
        return 1 + Math.max(leftHeight, rightHeight);

    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return dfs(root , LONG_MIN , LONG_MAX);
    }

    bool dfs (TreeNode* root , long minValue , long maxValue) {
        if(root == nullptr) {
            return true;
        }
        
        if(!(root -> val > minValue && root -> val < maxValue)){
            return false;
        }

        return dfs(root -> left , minValue , root -> val) && dfs(root -> right , root -> val , maxValue);
    }
};
